### PR TITLE
Replace debugLog in propagation logging

### PR DESCRIPTION
**Pull request by paulgessinger**
@HadrienG2, this is the branch where I got started with replacing the `debugLog` setup with our regular logging macros. This requires having logger instances in places that we didn't before, so some changes are required.

### CODE REVIEW

**Comment by paulgessinger** [comment]
The branch is a bit behind, unfortunately.

**Comment by HadrienG2** [comment]
That certainly looks promising!

**Comment by HadrienG2** [comment]
Once it builds, I can cross-check how much impact it has on build performance. Unlike with Measurement, it's a bit hard for me to precisely quantify how much impact removing `debugLog` would have because the compile-time overhead is spread out over many different std::function constructor instantiations.

**Comment by asalzburger** [comment]
> Once it builds, I can cross-check how much impact it has on build performance. Unlike with Measurement, it's a bit hard for me to precisely quantify how much impact removing `debugLog` would have because the compile-time overhead is spread out over many different `std::function` constructor instantiations.

Sounds good. If we have a good argument (which you seem to have) to remove the debugLog, then let's do it.

**Comment by paulgessinger** [comment]
Cool. We never really followed through and take this all the way. Combined with the fact that it makes debugging harder sometimes because its output is not easily correlated with other output and the computer time hit just makes it harder to justify.

**Comment by paulgessinger** [comment]
One consequence of using the current logger infrastructure in `Navigator` and `Propagator` is that they both become non-copyable (as `Logger` uses unique pointers, and it's convention to have `std::unique_ptr<Logger>` anyway). In the existing code this can be accommodated, but at some point we may want to revisit the forced consequence: `Logger` -> non-copyable.

**Comment by paulgessinger** [comment]
This builds on my machine now, let's see what the CI says.

**Comment by HadrienG2** [comment]
From a quick test run of the PropagationTests build, the benefit is measurable, but much smaller than I hoped for. We go from a max-RSS of 2.73GB on master to 2.66GB on this branch, so about -100MB.

Let's see if I can understand why...

**Comment by HadrienG2** [comment]
So, from a quick profiling run...

- The existing changes were quite already effective at saving compilation times in affected areas (e.g. `Acts::covariance_curvilinear<Acts::RiddersPropagator>` went trom 0.8s to 0.4s). So it may be the case that we've encountered a first example of a situation where using time profiles as a proxy for lack of memory profiles is misleading us.
- I can indeed see that the branch is not complete yet, as some debugLog usage is still visible in e.g. Propagator.ipp.

**Comment by paulgessinger** [comment]
Really, in `Propagator.ipp`? I thought there I had replaced them all. There's some replication here in smaller components which implement their own `debugLog`. (one reason more to get rid of it entirely, if it's not even centralized in the propagator).

**Comment by HadrienG2** [comment]
@paulgessinger I think you do not fully grasp the debugLog code duplication situation yet. Last time I checked, there were 8 copies of that function in the Acts codebase ;)

**Comment by HadrienG2** [comment]
Yup, you still have quite a few of these in your branch:

```
hadrien@spartan-graloufotron:~/Bureau/IJCLab/Programmation/acts> git checkout prop_logging 
Basculement sur la branche 'prop_logging'
Votre branche est à jour avec 'paulgessinger/prop_logging'.
hadrien@spartan-graloufotron:~/Bureau/IJCLab/Programmation/acts> rg --glob '!spack-build' -i 'debugLog'
Core/include/Acts/Propagator/Navigator.hpp
667:        debugLog(state, [&] {
1164:  void debugLog(propagator_state_t& state,

Core/include/Acts/Propagator/detail/LoopProtection.hpp
46:          debugLog(state, [&] {
70:  void debugLog(propagator_state_t& state,

Core/include/Acts/Propagator/Propagator.hpp
411:  void debugLog(propagator_state_t& state,

Core/include/Acts/Propagator/DirectNavigator.hpp
131:    debugLog(state, [&] { return std::string("Entering navigator::status."); });
136:    debugLog(state, [&] {
153:        debugLog(state, [&] {
163:          debugLog(state, [&] {
172:        debugLog(state, [&] {
192:    debugLog(state, [&] { return std::string("Entering navigator::target."); });
197:    debugLog(state, [&] {
211:        debugLog(state, [&] {
220:        debugLog(state, [&] {
234:        debugLog(state,
254:  void debugLog(propagator_state_t& state,

Core/include/Acts/Propagator/MaterialInteractor.hpp
131:        debugLog(state, [=] {
248:  void debugLog(propagator_state_t& state,

Core/include/Acts/Propagator/Propagator.ipp
18:  debugLog(state, [&] { return std::string("Entering propagation."); });
33:    debugLog(state, [&] { return std::string("Starting stepping loop."); });
42:        debugLog(state, [&] {
50:        debugLog(state, [&] {
74:    debugLog(state, [&] { return std::string("Terminated with failure."); });
79:  debugLog(state, [&] { return std::string("Stepping loop done."); });
223:void Acts::Propagator<S, N>::debugLog(

Core/include/Acts/Propagator/SurfaceCollector.hpp
103:      debugLog(state, [&] {
131:  void debugLog(propagator_state_t& state,

Core/include/Acts/Propagator/VolumeCollector.hpp
103:      debugLog(state, [&] {
131:  void debugLog(propagator_state_t& state,

Core/include/Acts/Propagator/StandardAborters.hpp
51:void targetDebugLog(propagator_state_t& state, const std::string& status,
89:      targetDebugLog(state, "x", [&] {
97:      targetDebugLog(state, "o", [&] {
144:      targetDebugLog(state, "x", [&] {
165:      targetDebugLog(state, "x", [&] {
173:      targetDebugLog(state, "x", [&] {
191:      targetDebugLog(state, "o", [&] {

Core/include/Acts/Material/MaterialCollector.hpp
69:      debugLog(state, [&] {
86:            debugLog(state, [&] {
94:            debugLog(state, [&] {
101:            debugLog(state, [&] {
109:            debugLog(state, [&] {
115:          debugLog(state, [&] {
128:          debugLog(state, [&] {
172:  void debugLog(propagator_state_t& state,
```

File: Core/include/Acts/Utilities/Logger.hpp [hunk 1]
```diff
@@ -521,6 +521,13 @@ class Logger {
                                         std::placeholders::_1));
   }
 
+  template <typename Callable>
+  void log(const Logging::Level& lvl, Callable&& callable) const {
+    if (doPrint(lvl)) {
+      callable(log(lvl));
+    }
+  }
+
```
Comment (paulgessinger) [inline Core/include/Acts/Utilities/Logger.hpp:530]: I added this as a replacement for one use-case of `debugLog`, which is to easily group logging code that only gets executed if the level is met. Do you think this has similar build-perf implications as `debugLog` @HadrienG2? It doesn't use `std::function`. If necessary, I think I can get rid of this as well.

**Comment by paulgessinger** [comment]
I moved the logger instance for Propagator/Navigator to the propagation options using only the new `LoggerWrapper`. This way, I could remove the `std::move`s that I had to add since Propagator/Navigator became non-copyable. This is better I think. I can use this now to get rid of `debugLog` in the actors.

File: Core/include/Acts/Utilities/Logger.hpp (hunk 1 above)
Comment (HadrienG2) [reply Core/include/Acts/Utilities/Logger.hpp:530]: It should have much less build-perf implications:

- `doPrint(lvl)` and `log(lvl)` are not templated on `Callable`, so they are only processed once by the compiler. The only thing that the compiler needs to process on every call is the lambda callable itself and a tiny "if this then that" code snippet that uses it.
- This is used much less frequently than `debugLog` was, so the compiler doesn't have that many lambdas to process.

**Comment by asalzburger** [comment]
Great work so far, guys!

**Comment by paulgessinger** [comment]
Ok, so I reverted the changes to `Propagator` and `Navigator` that added logger instances to them. A `LoggerWrapper` instance is now part of the propagator, KF and CKF options, and all logging aware code accesses it either from there, or is passed it. As kind of a side-effect, the log level can now be set for each individual propagation.

This change now removes some details from the debug output, which was added in debug log (stuff like the current volume). This could be added back in, e.g. by providing special logging macros or functions for the Navigator/Propagator, that don't use an `std::function` for delayed execution. @asalzburger, do you have thoughts here?

File: Core/CMakeLists.txt [hunk 2]
```diff
@@ -39,7 +39,6 @@ install(
 
 # target source files are added separately
 add_subdirectory(src/EventData)
-add_subdirectory(src/Fitter)
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Core/CMakeLists.txt:42]: I see, cpp files were removed.

File: Core/src/Utilities/Logger.cpp [hunk 3]
```diff
@@ -8,8 +8,50 @@
 
 #include "Acts/Utilities/Logger.hpp"
 
+#include <cassert>
+
 namespace Acts {
 
+LoggerWrapper::LoggerWrapper(const Logger& logger) : m_logger(&logger) {}
+
+bool LoggerWrapper::doPrint(const Logging::Level& lvl) const {
```
Comment (asalzburger) [inline Core/src/Utilities/Logger.cpp:17]: Did you write a test for this? Codecov does report non-coverage here.

**Comment by asalzburger** [comment]
I overwrite as it increases code coverage slightly.

**Pull request merged**